### PR TITLE
Delete existing tokens by scope intersection

### LINKED ISSUE #61: Ensure tokens with intersecting scopes are properly handled in the new schema cache
Make sure that cache writes/reads behave properly when scopes intersect

### PR DESCRIPTION
**Pull request by iambmelt**
Closes #61

### CODE REVIEW

**Comment by iambmelt** [comment]
TODO: Add unit tests to cover this

File: common/src/main/java/com/microsoft/identity/common/internal/cache/MsalOAuth2TokenCache.java [hunk 1]
```diff
@@ -60,9 +66,52 @@ private void saveAccessToken(
             final AuthorizationRequest request,
             final TokenResponse response) {
         final AccessToken accessToken = mAccountCredentialAdapter.createAccessToken(oAuth2Strategy, request, response);
+        deleteAccessTokensWithIntersectingScopes(accessToken);
```
Comment (heidijinxujia) [inline common/src/main/java/com/microsoft/identity/common/internal/cache/MsalOAuth2TokenCache.java]: I'm probably lack of context here. Why do we need to delete the access token of intersecting scope before saving the access token?

File: common/src/androidTest/java/com/microsoft/identity/common/MsalOAuth2TokenCacheTest.java [hunk 2]
```diff
@@ -0,0 +1,231 @@
...
+import static org.mockito.Mockito.when;
+
+@RunWith(AndroidJUnit4.class)
+public class MsalOAuth2TokenCacheTest extends AndroidSecretKeyEnabledHelper {
+
+    private static final String UNIQUE_ID = "29f3807a-4fb0-42f2-a44a-236aa0cb3f97.0287f963-2d72-4363-9e3a-5705c5b0f031";
+    private static final String ENVIRONMENT = "login.microsoftonline.com";
+    private static final String CLIENT_ID = "0287f963-2d72-4363-9e3a-5705c5b0f031";
+    private static final String TARGET = "user.read user.write https://graph.windows.net";
+    private static final String REALM = "3c62ac97-29eb-4aed-a3c8-add0298508d";
```
Comment (shoatman) [inline common/src/androidTest/java/com/microsoft/identity/common/MsalOAuth2TokenCacheTest.java]: Are these tenants?  If so we may want to call them tenants.

Comment (shoatman) [inline common/src/androidTest/java/com/microsoft/identity/common/MsalOAuth2TokenCacheTest.java]: I'd be ok leaving a comment here and other appropriate places that Realm equals tenant in the case of AAD.

File: common/src/main/java/com/microsoft/identity/common/internal/cache/MsalOAuth2TokenCache.java (hunk 1 above)
Comment (shoatman) [reply common/src/main/java/com/microsoft/identity/common/internal/cache/MsalOAuth2TokenCache.java]: The v2 endpoint always returns all of the user consented scopes associated with a particular resource.  As a result if the scopes intersect then the existing access token can be replaced.... or simply deleted.

**Comment by iambmelt** [comment]
@shoatman We should see what iOS is doing on this point -- I was operating under the impression that _all_ scopes would saved in cache and in the key. @oldalton Is this consistent with your implementation?

If necessary, I can modify the logic to not include `openid` and `profile` in it's intersection criteria.

File: common/src/androidTest/java/com/microsoft/identity/common/MsalOAuth2TokenCacheTest.java (hunk 2 above)
Comment (iambmelt) [reply common/src/androidTest/java/com/microsoft/identity/common/MsalOAuth2TokenCacheTest.java]: This is javadoc'd inside of `Account` -- I'll add an additional note here

**Comment by oldalton** [comment]
@iambmelt In iOS we're saving based on the scopes in the response. There's no special criteria to remove openid or profile scopes, so we're saving all scopes returned by the service. Since in token lookups we're matching a subset, there's no harm in saving more scopes than necessary.

**Comment by shoatman** [comment]
@iambmelt @oldalton I double checked.  The server does not return openid or profile scopes in the scopes response parameter even when they have been granted.

**Comment by oldalton** [comment]
@shoatman Thanks!

**Comment by iambmelt** [comment]
@shoatman Comments addressed. Feel free to have another look

**Comment by shoatman** [comment]
@iambmelt Looks good

**Pull request merged**